doc/tutorial.rst

```diff
@@ -0,0 +1,317 @@
+
+.. code-block:: sh
+
+   $ chmod +x testsuite.py
+   $ ./testsuite.py
+   INFO     Found 0 tests
+   INFO     Summary:
+   _  <no test result>
+
+That makes sense: we have an empty testsuite, so running it actually executes
+no test.
+
+
+.. _tutorial_creating_test_driver:
+
+Creating a test driver
+----------------------
+
+Most testcases will check the behavior of artithmetic computations, so we have
+an obvious first driver to write: it will spawn a ``bc`` process, passing a
+file that contains the arithmetic expression to evaluate to it and check that
+the output is as expected. Testcases using this driver just need to provide the
+expression input file and an expected output file.
+
+Creating a test driver is as simple as creating a Python class that derives
+from ``e3.testsuite.drivers.TestDriver``. However, its API is quite crude, so
+we will study it :ref:`later <api_test_driver>`. Instead, let's use
+``e3.testsuite.drivers.diff.DiffTestDriver`` instead: that class conveniently
+provides the framework to spawn subprocesses and check their outputs against
+baselines, i.e. exactly what we want to do here.
+
+Add the following class to ``testsuite.py``:
+
+.. code-block:: python
+
+   from e3.testsuite.driver.diff import DiffTestDriver
+
+
+   class ComputationDriver(DiffTestDriver):
+       """Driver to run a computation through "bc" and check its output.
+
+       This passes the "input.bc" to "bc" and check its output against the
+       "test.out" baseline file.
+       """
+
+       def run(self):
+           self.shell(["bc", "input.bc"])
+
+The only mandatory thing to do for ``ClassicTestDriver`` concrete subclasses
+(``DiffTestDriver`` is an abstract subclass) is to override the ``run`` method.
+The role of this method is to do whatever actions the test driver is supposed
+to do in order for testcases to exercize the tested piece of software: compile
+software, prepare input files, run processes, and so on.
+
+Here, the only thing we need to do is to actually run the ``bc`` program on our
+input file. ``shell`` is a method inherited from ``ClassicTestDriver`` acting
+as a wrapper around Python's ``subprocess`` standard library module. This
+wrapper spawns a subprocess with a empty standard input, returns its exit
+code and captured output (mix of standard output/error) and . While the only
+mandatory argument is a list of strings for the command-line to run, optional
+arguments control how to spawn this subprocess and use its result, for
+instance:
+
+* ``cwd`` controls the working directory for the subprocess.
+* ``env`` allows to control environment variables passed to the subprocess. By
+  default: leave the testsuite environment variables unchanged.
+* ``catch_error``: whether to consider non-zero exit code as a test failure
+  (true by default).
+* ``analyze_output``: whether to append the subprocess' output to
+  ``self.output``, the attribute used to store the output compared against the
+  test baseline (true by default).
```

Comment:
Unclear. In fact, I don't understand what the "output" is when this is set to false. Only the exit code?

Reply:
No: `self.output` is a string (or more precisely a Log instance… which is introduced later) that is used to compute a diff when the test execution is completed (diff against the test baseline). `analyze_output` controls whether `self.shell` appends the subrocess’ output (stdout/stderr mix, described in the § above) to `self.output`… i.e. whether to consider its output for the diff. I need to think about a better way to describe this without giving too much details. Maybe I should just skip the description and say “we’ll see later”.

Reply:
Indeed, I understand that part better after having read the next sections. This is a bit advanced indeed. I would:
- mention `self.output` earlier, saying that the test driver can write in it on its own
- remove the explanation about `self.output`.
Hopefully the reader won't have forgotten what it is, or will go back to refresh his memory.

Reply:
Thanks, I’ll do that!